[Serialization] Do less work checking if a value can be deserialized.

jrose-apple opened the pull request:
Previously we recorded the canonical type of the declaration and made sure we could deserialize that, but that's a lot of extra work building up intermediate types that we mostly don't need. Instead, record smaller types that represent the possible points of failure—right now, just the nominal types that are referenced by the value (function, variable/constant, subscript, or initializer). I chose to use types instead of declarations here because types can potentially encode more complicated constraints later (such as generic types checking that their arguments still conform).

This gains us back 20% of type-checking time on a compile-time microbenchmark: `let _ = [1, 2]`. I expect the effect is less dramatic the more expressions you have, since we only need to deserialize things once.

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test macOS

DougGregor commented on lib/Serialization/Deserialization.cpp:
```diff
@@ -2810,37 +2818,42 @@ ModuleFile::getDeclChecked(DeclID DID, Optional<DeclContext *> ForcedContext) {
+          nameAndDependencyIDs.slice(numNameComponentsBiased);
+    } else {
+      name = baseName;
+      nameAndDependencyIDs = nameAndDependencyIDs.drop_front();
```
The change of meaning of `nameAndDependencyIDs` here really threw me. The code is correct, but kinda hard to read. What about just defining

    auto dependencyIDs = nameAndDependencyIDs.slice(numNameComponentsBiased);

?

DougGregor commented on lib/Serialization/Serialization.cpp:
```diff
@@ -2315,6 +2315,15 @@ void Serializer::writeForeignErrorConvention(const ForeignErrorConvention &fec){
+  SmallVector<Type, 2> result;
+  ty.visit([&](CanType next) {
+    if (auto *nominal = next->getAnyNominal())
+      result.push_back(nominal->getDeclaredType());
```
How about `getDeclaredInterfaceType` here, so we don't have any archetypes in there that would force deserialization of a generic environment?

jrose-apple replied:
getDeclaredType should already be using UnboundGenericType, no? I wanted the absolute smallest type.

jrose-apple replied:
…but that probably doesn't make sense for nested types. Okay, getDeclaredInterfaceType it is.

DougGregor commented on lib/Serialization/Serialization.cpp:
```diff
@@ -2315,6 +2315,15 @@ void Serializer::writeForeignErrorConvention(const ForeignErrorConvention &fec){
+    if (auto *nominal = next->getAnyNominal())
+      result.push_back(nominal->getDeclaredType());
+  });
+  return result;
```
Should you unique the types in `result`?

jrose-apple replied:
Ah, good idea.

DougGregor reviewed (approved):
Some minor suggestions for improvement.

jrose-apple committed:
[Serialization] Do less work checking if a value can be deserialized.

Previously we recorded the canonical type of the declaration and made
sure we could deserialize that, but that's a lot of extra work
building up intermediate types that we mostly don't need. Instead,
record smaller types that represent the possible points of failure---
right now, just the nominal types that are referenced by the value
(function, variable/constant, subscript, or initializer). I chose to
use types instead of declarations here because types can potentially
encode more complicated constraints later (such as generic types
checking that their arguments still conform).

This gains us back 20% of type-checking time on a compile-time
microbenchmark: `let _ = [1, 2]`. I expect the effect is less dramatic
the more expressions you have, since we only need to deserialize
things once.

jrose-apple commented:
[macOS full test on prior revision](https://ci.swift.org/job/swift-PR-osx/7695/)

@swift-ci Please smoke test

DougGregor reviewed (approved):
LGTM, thank you!

jrose-apple commented:
```
[ RUN      ] TestDateInterval.test_EncodingRoundTrip_JSON
stdout>>> check failed at /Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-test/branch-master/swift/test/stdlib/Inputs/FoundationTestsShared.swift, line 19
stdout>>> expected: 0000-12-30 00:00:00 +0000 to 2017-05-17 01:04:59 +0000 (of type Foundation.DateInterval)
stdout>>> actual: 0000-12-30 00:00:00 +0000 to 2017-05-17 01:04:59 +0000 (of type Foundation.DateInterval)
stdout>>> Decoded DateInterval <0000-12-30 00:00:00 +0000 to 2017-05-17 01:04:59 +0000> not equal to original <0000-12-30 00:00:00 +0000 to 2017-05-17 01:04:59 +0000>
stdout>>> check failed at /Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-test/branch-master/swift/test/stdlib/Inputs/FoundationTestsShared.swift, line 19
stdout>>> expected: 2017-05-17 01:04:59 +0000 to 4001-01-01 00:00:00 +0000 (of type Foundation.DateInterval)
stdout>>> actual: 2017-05-17 01:04:59 +0000 to 4001-01-01 00:00:00 +0000 (of type Foundation.DateInterval)
stdout>>> Decoded DateInterval <2017-05-17 01:04:59 +0000 to 4001-01-01 00:00:00 +0000> not equal to original <2017-05-17 01:04:59 +0000 to 4001-01-01 00:00:00 +0000>
[     FAIL ] TestDateInterval.test_EncodingRoundTrip_JSON
```

@itaiferber?

jrose-apple commented:
@swift-ci Please smoke test

jrose-apple commented:
@swift-ci Please smoke test macOS

itaiferber commented:
@jrose-apple Strange... I'll disable the test.

itaiferber commented:
#9682

itaiferber commented:
@jrose-apple Should be clear now.

jrose-apple commented:
@swift-ci Please smoke test macOS
